alibi/explainers/anchor_text.py

```diff
@@ -421,69 +402,948 @@ def perturb_sentence(self, present: tuple, n: int, sample_proba: float = 0.5,
+
+            # select top n tokens from each distribution
+            top_k = tf.math.top_k(logits_mask, top_n)
+            top_k_logits, top_k_tokens = top_k.values, top_k.indices
+            top_k_logits = (top_k_logits / temperature) if use_proba else (top_k_logits * 0)
+
+            # sample `num_samples` instance for the current mask template
+            for j in range(mult_factor + int(i < reminder)):
+                # Compute the buffer index
+                idx = i * mult_factor + j + min(i, reminder)
+
+                # Sample indices
+                ids_k = tf.reshape(tf.random.categorical(top_k_logits, 1), shape=-1)
+
+                # Set the unmasked tokens and for the masked one and replace them with the samples drawn
+                sampled_tokens[idx] = tokens[i]
+                sampled_tokens[idx, cols] = tf.gather(top_k_tokens, ids_k, batch_dims=1)
+
+            # Add the original binary mask which marks the beginning of a masked
+            # word, as is needed for the anchor algorithm (backend stuff)
+            idx, offset = i * mult_factor, min(i, reminder)
+            sampled_data[idx + offset:idx + mult_factor + offset + (i < reminder)] = data[i]
+
+        # Check that there are not masked tokens left
+        assert np.all(sampled_tokens != self.model.mask_token)
+        assert np.all(np.any(sampled_tokens != 0, axis=1))
+        return sampled_tokens, sampled_data
+
+    def _perturb_instance_ar(self,
+                             num_samples: int,
+                             raw: np.ndarray,
+                             data: np.ndarray,
+                             top_n: int = 100,
+                             batch_size: int = 32,
+                             temperature: float = 1.0,
+                             use_proba: bool = False,
+                             **kwargs) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Perturb the instances in an autoregressive fashion (sequential).
+
+        ----------
+        num_samples
+            Number of samples to be generates.
+        raw
+            Array of mask templates. Has `mask_templates` rows.
+        data
+            Binary array having 0 where the tokens are masked. Has `mask_templates` rows.
+        top_n:
+            Use the top n words when sampling.
+        batch_size_lm:
+            Batch size used for language model.
+        temperature
+            Sample weight hyper-parameter.
+        use_proba
+            Bool whether to sample according to the predicted words distribution.
+
+        Returns
+        -------
+        sampled_tokens
+            Tensor containing the ids of the sampled tokens.
+            Has `num_samples` rows.
+        sampled_data
+            Binary array havin 0 where the tokens were masked.
+            Has `num_samples` rows.
+        """
+        # number of samples to generate per mask template
+        assert num_samples == raw.shape[0]
```

Comment:
same comment as for other asserts, can this fail?

Reply:
The assertion should not be triggered. It is just for testing purposes.